controllers/apps/operations/switchover.go

```diff
@@ -0,0 +1,307 @@
+
+This program is distributed in the hope that it will be useful
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+package operations
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1alpha1 "github.com/apecloud/kubeblocks/apis/apps/v1alpha1"
+	componentutil "github.com/apecloud/kubeblocks/controllers/apps/components/util"
+	"github.com/apecloud/kubeblocks/internal/constant"
+	intctrlutil "github.com/apecloud/kubeblocks/internal/controllerutil"
+)
+
+type switchoverOpsHandler struct{}
+
+var _ OpsHandler = switchoverOpsHandler{}
+
+// SwitchoverMessage is the OpsRequest.Status.Condition.Message for switchover.
+type SwitchoverMessage struct {
+	appsv1alpha1.Switchover
+	OldPrimaryOrLeader string
+	Cluster            string
+}
+
+func init() {
+	switchoverBehaviour := OpsBehaviour{
+		FromClusterPhases:                  appsv1alpha1.GetClusterUpRunningPhases(),
+		ToClusterPhase:                     appsv1alpha1.SpecReconcilingClusterPhase,
+		OpsHandler:                         switchoverOpsHandler{},
+		MaintainClusterPhaseBySelf:         true,
+		ProcessingReasonInClusterCondition: ProcessingReasonVersionSwitchovering,
+	}
+
+	opsMgr := GetOpsManager()
+	opsMgr.RegisterOps(appsv1alpha1.SwitchoverType, switchoverBehaviour)
+}
+
+// ActionStartedCondition the started condition when handle the switchover request.
+func (r switchoverOpsHandler) ActionStartedCondition(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource) (*metav1.Condition, error) {
+	switchoverMessageMap := make(map[string]SwitchoverMessage)
+	for _, switchover := range opsRes.OpsRequest.Spec.SwitchoverList {
+		pod, err := getPrimaryOrLeaderPod(reqCtx.Ctx, cli, *opsRes.Cluster, switchover.ComponentName, opsRes.Cluster.Spec.GetComponentDefRefName(switchover.ComponentName))
+		if err != nil {
+			return nil, err
+		}
+		switchoverMessageMap[switchover.ComponentName] = SwitchoverMessage{
+			Switchover:         switchover,
+			OldPrimaryOrLeader: pod.Name,
+			Cluster:            opsRes.Cluster.Name,
+		}
+	}
+	msg, err := json.Marshal(switchoverMessageMap)
+	if err != nil {
+		return nil, err
+	}
+	return appsv1alpha1.NewSwitchoveringCondition(opsRes.Cluster.Generation, string(msg)), nil
+}
+
+// Action restarts components by updating StatefulSet.
+func (r switchoverOpsHandler) Action(reqCtx intctrlutil.RequestCtx, cli client.Client, opsRes *OpsResource) error {
+	if opsRes.OpsRequest.Status.StartTimestamp.IsZero() {
+		return errors.New("status.startTimestamp can not be null")
+	}
```

Comment:
Code is redundant, startTimestamp will not be empty before exec the Action function